### DISCUSSION TITLE
Uploading Vector data to Qdrant Server

### DISCUSSION THREAD

**Discussion by harshakhmk** [qdrant/qdrant-client#444] [Q&A]
I am using Qdrant Client to upload an array of vector data where each vector length is 785, how to add this bulk of vectors to Qdrant Server using Qdrant Client python package

**Comment by joein** [comment]
Hi @harshakhmk 

You can either construct batches on your own and use [upsert API](https://qdrant.tech/documentation/concepts/points/?selector=aHRtbCA%2BIGJvZHkgPiBkaXY6bnRoLW9mLXR5cGUoMSkgPiBzZWN0aW9uID4gZGl2ID4gZGl2ID4gZGl2Om50aC1vZi10eXBlKDIpID4gYXJ0aWNsZSA%2BIGgyOm50aC1vZi10eXBlKDMp#upload-points)

or you can rely on qdrant-client and use methods like [upload_records](https://qdrant.tech/documentation/tutorials/search-beginners/?selector=aHRtbCA%2BIGJvZHkgPiBkaXY6bnRoLW9mLXR5cGUoMSkgPiBzZWN0aW9uID4gZGl2ID4gZGl2ID4gZGl2Om50aC1vZi10eXBlKDIpID4gYXJ0aWNsZSA%2BIGgyOm50aC1vZi10eXBlKDYp#5-upload-data-to-collection) or [upload_collection](https://github.com/qdrant/qdrant-client/blob/6d019e67a133bea87a96bce388f5b901cdae1287/qdrant_client/qdrant_client.py#L1812C7-L1812C7)

`upload_collection` and `upload_records` methods also have `parallel` param, which allows to seamlessly parallelise (via multiprocessing) uploading

**Comment by joein** [comment]
Hi @harshakhmk, does it help?

**Comment by harshakhmk** [comment]
Hi, I tried upload _records and I was getting pydantic validation errors
```
pydantic_core._pydantic_core.ValidationError: 1 Validation Error for Record payload
Input should be a valid dictionary [ type=dict_type, input_value = array([ 0,0,0... 0,0]), input_type=ndarray]
For further information, visit https://errors.pydantic.dev/2.5/v/dict_type

```

**Comment by joein** [comment]
Could you share a code sample to reproduce?
How do you make an `upload_records` call?

**Comment by harshakhmk** [comment]
```

import pandas as pd
df=pd.read_csv("test.csv")
Vectors1=[]
for idx in range (len(df)):
   item=df.iloc[idx]
   vectors1.append(item)



client.upload_records(
collection_name="dataset",
records=[
models. Record(
id=idx , vector=doc.tolist(), payload=doc
)
for idx, doc in enumerate(vectors1)
],
)

**Comment by joein** [comment]
`payload=doc` is the source of error I believe

it should be a dict, but you are uploading a numpy array there 

you need to do something like `payload={"vector": doc.tolist()}`

btw, are you sure you want to store vector in the payload?

**Comment by harshakhmk** [comment]
Yes, that was the issue, I tried other way to insert data using upsert function, since it cannot insert all vectors at once, I had wrapped insertion of data to server inside a function and called this function for each vector, I had observed huge difference in adding whole collection to database with those two approaches
upload_records is much faster

**Comment by harshakhmk** [comment]
But, what will be the batch size with upload_records as we are not explicitly defined

**Comment by joein** [comment]
`upsert` function presumes manual batching

`upload_records` contains a default parameter `batch_size: int = 64`

To speed up things even more you can set `parallel: int = <number of processes>` which will spread the load across CPU processes

One thing to be aware of while using `upload_records`

If you don't provide ids explicitly, it will use consequent int ids (like `0` for the first record, `1` for the second, etc.)

If you call it twice - then on the second time it will  overwrite the points which were written before

To sum up: if you provide ids explicitly - never mind, if you don't provide ids explicitly - be careful while calling to `upload_records` the second time
